include/sys/emmc.h

```diff
@@ -0,0 +1,279 @@
+  EMMC_CMD_BUSTEST_R = 14,
+  EMMC_CMD_GO_INACTIVE_STATE = 15,
+  EMMC_CMD_BUSTEST_W = 19,
+  EMMC_CMD_SET_BLOCKLEN = 16,
+  EMMC_CMD_READ_BLOCK = 17,
+  EMMC_CMD_READ_MULTIPLE_BLOCKS = 18,
+  EMMC_CMD_WRITE_DAT_UNTIL_STOP = 20,
+  EMMC_CMD_SET_BLOCK_COUNT = 23,
+  EMMC_CMD_WRITE_BLOCK = 24,
+  EMMC_CMD_WRITE_MULTIPLE_BLOCKS = 25,
+  EMMC_CMD_PROGRAM_CID = 26,
+  EMMC_CMD_PROGRAM_CSD = 27,
+  EMMC_CMD_SET_WRITE_PROT = 28,
+  EMMC_CMD_CLR_WRITE_PROT = 29,
+  EMMC_CMD_SEND_WRITE_PROT = 30,
+  EMMC_CMD_SEND_WRITE_PROT_TYPE = 31,
+  EMMC_CMD_ERASE_GROUP_START = 35,
+  EMMC_CMD_ERASE_GROUP_END = 36,
+  EMMC_CMD_ERASE = 38,
+  EMMC_CMD_FAST_IO = 39,
+  EMMC_CMD_GO_IRQ_STATE = 40,
+  EMMC_CMD_LOCK_UNLOCK = 42,
+  EMMC_CMD_APP_CMD = 55,
+  EMMC_CMD_GEN_CMD = 56,
+} emmc_command_t;
+
+#define EMMC_I_DATA_DONE 0x01
+#define EMMC_I_WRITE_READY 0x02
+#define EMMC_I_READ_READY 0x04
+
+typedef uint64_t emmc_wait_flags_t;
+
+/* At most one of these */
+#define EMMC_APP_SUSPEND 0x01
+#define EMMC_APP_RESUME 0x02
+#define EMMC_APP_ABORT 0x03
+/* Any subset of these */
+#define EMMC_APP_DATA 0x04
+#define EMMC_APP_CHKIDX 0x08
+#define EMMC_APP_CHKCRC 0x10
+/* At most one of these */
+#define EMMC_APP_RESP136 0x20
+#define EMMC_APP_RESP48 0x40
+
+#define EMMC_APP_NEEDS_CMDTYPE 0x03
+#define EMMC_APP_NEEDS_RESPTYPE 0x60
+
+typedef uint32_t emmc_app_flags_t;
+
+/* For a detailed explanation on semantics refer to the comments above
+ * respective wrappers */
+typedef emmc_result_t (*emmc_send_cmd_t)(device_t *dev, emmc_command_t cmd,
+                                         uint32_t arg1, uint32_t arg2,
+                                         emmc_response_t *res);
+typedef emmc_result_t (*emmc_send_app_cmd_t)(device_t *dev, uint32_t cmd,
+                                             uint32_t arg1, uint32_t arg2,
+                                             emmc_resp_type_t exp_rt,
+                                             emmc_app_flags_t app_flags,
+                                             emmc_response_t *res);
+typedef int (*emmc_wait_t)(device_t *dev, emmc_wait_flags_t wflags);
+typedef emmc_result_t (*emmc_read_dat_t)(device_t *dev, void *buf, size_t len,
+                                         size_t *read);
+typedef emmc_result_t (*emmc_write_dat_t)(device_t *dev, const void *buf,
+                                          size_t len, size_t *wrote);
+typedef int (*emmc_get_prop_t)(device_t *dev, uint32_t id, uint64_t *val);
+typedef int (*emmc_set_prop_t)(device_t *dev, uint32_t id, uint64_t val);
+
+#define EMMC_PROP_R_MODE 1
+#define EMMC_PROP_RW_BLKSIZE 2
+#define EMMC_PROP_RW_BLKCNT 3
+#define EMMC_PROP_R_INT_FLAGS 4
+
+typedef struct emmc_methods {
+  emmc_send_cmd_t send_cmd;
+  emmc_send_app_cmd_t send_app_cmd;
+  emmc_wait_t wait;
+  emmc_read_dat_t read;
+  emmc_write_dat_t write;
+  emmc_get_prop_t get_prop;
+  emmc_set_prop_t set_prop;
+} emmc_methods_t;
+
+typedef struct emmc_device {
+  const uint64_t cid[2];
```

Comment:
Maybe a macro defining (and describing) these array sizes?

Reply:
It would be quite unnecessary. These were just high and low bytes of an 128-bit value. Anyway, the whole response thing is different now, and hopefully the new macros defined for accessing response data make it clear what's going on here.